File: crates/bindings-typescript/src/lib/option.ts
import { AlgebraicType } from './algebraic_type';

export type OptionAlgebraicType<T extends AlgebraicType = AlgebraicType> = {
  tag: 'Sum';
  value: {
    variants: [
      { name: 'some'; algebraicType: T },
      {
        name: 'none';
        algebraicType: { tag: 'Product'; value: { elements: [] } };
      },
    ];
  };
};

export const Option: {
  getAlgebraicType<T extends AlgebraicType = AlgebraicType>(
    innerType: T
  ): OptionAlgebraicType<T>;
} = {
  getAlgebraicType<T extends AlgebraicType = AlgebraicType>(
    innerType: T
  ): OptionAlgebraicType<T> {
    return AlgebraicType.Sum({
      variants: [
        { name: 'some', algebraicType: innerType },
        {
          name: 'none',
          algebraicType: AlgebraicType.Product({ elements: [] }),
        },
      ],
    });
  },
};

File: sdks/go/bsatn/option.go
package bsatn

// WriteOption writes an Option<T> value as a BSATN sum type.
// Some(value) = tag 0 + encoded value
// None = tag 1 (empty product)
func WriteOption[T Serializable](w Writer, v *T) {
	if v != nil {
		w.PutSumTag(0)
		(*v).WriteBsatn(w)
	} else {
		w.PutSumTag(1)
	}
}

// ReadOption reads a BSATN Option<T> using the provided read function.
func ReadOption[T any](r Reader, readFn func(Reader) (T, error)) (*T, error) {
	tag, err := r.GetSumTag()
	if err != nil {
		return nil, err
	}
	switch tag {
	case 0: // Some
		v, err := readFn(r)
		if err != nil {
			return nil, err
		}
		return &v, nil
	case 1: // None
		return nil, nil
	default:
		return nil, &ErrInvalidTag{Tag: tag, SumName: "Option"}
	}
}
